File: python/day1/main.py
from pathlib import Path
from argparse import ArgumentParser


def parse(file: Path):
    with file.open() as fp:
        while line := fp.readline():
            yield int(line.strip().replace("R", "").replace("L", "-"))


def solve1(data: list[int]):
    val = 50
    cnt = 0
    for n in data:
        val = (val + n) % 100
        if val == 0:
            cnt += 1
    return cnt


def solve2(data: list[int]):
    val = 50
    cnt = 0
    for n in data:
        if n > 0:
            cnt += (val + n) // 100 - val // 100
        elif n < 0:
            cnt += (val - 1) // 100 - (val + n - 1) // 100
        val = (val + n) % 100
    return cnt
    

parser = ArgumentParser()
parser.add_argument("input", type=Path)
args = parser.parse_args()

nums = list(parse(args.input))
print(solve1(nums))
print(solve2(nums))

File: rust/day1/src/main.rs
use common::{get_lines, Result};
use std::env;
use std::path::Path;

fn parse(file: &Path) -> Result<Vec<i64>> {
  let lines = get_lines(file)?;
  Ok(
    lines
      .iter()
      .map(|line| match line.chars().next().unwrap() {
        'L' => -line[1..].parse::<i64>().unwrap(),
        'R' => line[1..].parse::<i64>().unwrap(),
        _ => panic!("Invalid char {}", line),
      })
      .collect(),
  )
}

fn solve1(data: &[i64]) -> i64 {
  data
    .iter()
    .scan(50i64, |val, &n| {
      *val = (*val + n).rem_euclid(100);
      Some(*val)
    })
    .filter(|&v| v == 0)
    .count() as i64
}

fn solve2(data: &[i64]) -> i64 {
  data
    .iter()
    .scan(50i64, |val, &n| {
      let crossings = if n > 0 {
        (*val + n).div_euclid(100) - val.div_euclid(100)
      } else if n < 0 {
        (*val - 1).div_euclid(100) - (*val + n - 1).div_euclid(100)
      } else {
        0
      };
      *val = (*val + n).rem_euclid(100);
      Some(crossings)
    })
    .sum()
}

fn main() -> Result<()> {
  let input = env::args()
    .nth(1)
    .expect("usage: day1 <input-file>");
  let nums = parse(Path::new(&input))?;
  println!("{}", solve1(&nums));
  println!("{}", solve2(&nums));
  Ok(())
}
